src/main/java/picard/illumina/ExtractBarcodesProgram.java

```diff
@@ -0,0 +1,312 @@
+                }
+                return isValidPrefix;
+            }).collect(Collectors.toList());
+
+            if (readStructure.sampleBarcodes.length() != validBarcodeColumns.size()) {
+                messages.add("Expected " + readStructure.sampleBarcodes.length() + " valid barcode columns, but found " +
+                        String.join(",", validBarcodeColumns));
+            }
+
+            validBarcodeColumns.sort((s, t1) -> {
+                int lengthDiff =  s.length() - t1.length();
+                if(lengthDiff == 0) return s.compareTo(t1);
+                return lengthDiff;
+            });
+
+            Matcher matcher = Pattern.compile("^(.*)_\\d").matcher(validBarcodeColumns.get(0));
+
+            final String sequenceColumn;
+            boolean hasMultipleNumberedBarcodeColumns = matcher.matches();
+
+            // If there are multiple numbered barcode columns we extract the base column name (ie for BARCODE_1, BARCODE_2)
+            // the sequence column is BARCODE. In the case of BARCODE_PARAMS file the columns are BARCODE_SEQUENCE_1
+            // and BARCODE_SEQUENCE_2.
+            if (hasMultipleNumberedBarcodeColumns) {
+                sequenceColumn = matcher.group(1);
+            } else {
+                sequenceColumn = validBarcodeColumns.get(0);
+            }
+
+            final boolean hasBarcodeName = barcodesParser.hasColumn(BARCODE_NAME_COLUMN);
+            final boolean hasLibraryName = barcodesParser.hasColumn(LIBRARY_NAME_COLUMN);
+            final int numBarcodes = readStructure.sampleBarcodes.length();
+            final Set<String> barcodes = new HashSet<>();
+            for (final TabbedTextFileWithHeaderParser.Row row : barcodesParser) {
+                final String[] bcStrings = new String[numBarcodes];
+                int barcodeNum = 0;
+                for (final ReadDescriptor rd : readStructure.descriptors) {
+                    if (rd.type != ReadType.Barcode) {
+                        continue;
+                    }
+                    final String header = hasMultipleNumberedBarcodeColumns ? sequenceColumn + "_" + (1 + barcodeNum) : sequenceColumn;
+                    final String field = row.getField(header);
+                    if (field == null) {
+                        messages.add(String.format("Null barcode in column %s of row: %s", header, row.getCurrentLine()));
+                        bcStrings[barcodeNum] = "";
+                    } else {
+                        bcStrings[barcodeNum] = field;
+                    }
+                    ++barcodeNum;
+                }
+                final String bcStr = IlluminaUtil.barcodeSeqsToString(bcStrings);
+                // if the barcode is all Ns don't add it to metrics (we add noCallMetric separately)
+                if(bcStr.contains("N") || bcStr.contains("n")) continue;
```

Comment:
```suggestion
                if (bcStr.contains("N") || bcStr.contains("n")) {
                  continue;
                }
```
But, we hates 'continue;'